File: src/hooks/ipc/runtime.ts
import { invoke, isTauri } from '@tauri-apps/api/core';
import type { AppBuildInfo, RuntimeCapabilities, RuntimeCapabilityKey } from '../../types';
import { loadBrowserPreviewState } from '../../utils/browserPreviewSeed';

interface RustRuntimeCapabilities {
  device_detection: boolean;
  heuristic_diagnostic: boolean;
  ai_advisory: boolean;
  optional_cloud_ai: boolean;
  scan_engine: boolean;
  imaging_engine: boolean;
  results_browser: boolean;
  export_validation: boolean;
  export_engine: boolean;
  technical_logs: boolean;
  limited_capabilities?: string[];
}

interface RustAppBuildInfo {
  product_name: string;
  bundle_identifier: string;
  app_version: string;
  package_name: string;
  build_profile: string;
  operating_system: string;
  architecture: string;
  target_triple: string;
  tauri_runtime: string;
}

export async function fetchRuntimeCapabilities(): Promise<RuntimeCapabilities> {
  if (__ALLOW_BROWSER_PREVIEW__ && !isTauri()) {
    return (
      loadBrowserPreviewState()?.runtimeCapabilities ?? {
        deviceDetection: false,
        heuristicDiagnostic: false,
        aiAdvisory: false,
        optionalCloudAi: false,
        scanEngine: false,
        imagingEngine: false,
        resultsBrowser: true,
        exportValidation: false,
        exportEngine: false,
        technicalLogs: false,
        limitedCapabilities: [],
      }
    );
  }
  const c = await invoke<RustRuntimeCapabilities>('get_runtime_capabilities');
  return {
    deviceDetection: c.device_detection,
    heuristicDiagnostic: c.heuristic_diagnostic,
    aiAdvisory: c.ai_advisory,
    optionalCloudAi: c.optional_cloud_ai,
    scanEngine: c.scan_engine,
    imagingEngine: c.imaging_engine,
    resultsBrowser: c.results_browser,
    exportValidation: c.export_validation,
    exportEngine: c.export_engine,
    technicalLogs: c.technical_logs,
    limitedCapabilities: (c.limited_capabilities ?? []) as RuntimeCapabilityKey[],
  };
}

export async function fetchAppBuildInfo(): Promise<AppBuildInfo> {
  if (__ALLOW_BROWSER_PREVIEW__ && !isTauri()) {
    const buildInfo = loadBrowserPreviewState()?.buildInfo;
    return {
      productName: buildInfo?.productName ?? 'Recupere',
      bundleIdentifier: buildInfo?.bundleIdentifier ?? 'com.recupere.browser-preview',
      appVersion: buildInfo?.appVersion ?? __APP_VERSION__,
      packageName: buildInfo?.packageName ?? 'recupere',
      buildProfile: buildInfo?.buildProfile ?? 'browser-preview',
      operatingSystem: buildInfo?.operatingSystem ?? 'browser',
      architecture: buildInfo?.architecture ?? 'web',
      targetTriple: buildInfo?.targetTriple ?? 'browser-preview',
      tauriRuntime: buildInfo?.tauriRuntime ?? 'browser-preview',
    };
  }
  const info = await invoke<RustAppBuildInfo>('get_app_build_info');
  return {
    productName: info.product_name,
    bundleIdentifier: info.bundle_identifier,
    appVersion: info.app_version,
    packageName: info.package_name,
    buildProfile: info.build_profile,
    operatingSystem: info.operating_system,
    architecture: info.architecture,
    targetTriple: info.target_triple,
    tauriRuntime: info.tauri_runtime,
  };
}

File: src-tauri/src/commands/runtime.rs
// ============================================================================
// Récupère — Runtime introspection commands
// ============================================================================
// Surface host runtime capabilities (which engines are available) and build
// identity (product name, bundle id, target triple) to the front-end. The
// computation lives here; `commands/mod.rs` re-exports `runtime_capabilities`
// and `app_build_info` via `pub(super) use` so that callers in neighbouring
// modules (scan, ai, support-bundle builder, tests) keep the bare name.
// ============================================================================

use std::env;

use crate::types::{AppBuildInfo, RuntimeCapabilities};

pub(crate) const APP_PRODUCT_NAME: &str = "Récupère";
pub(crate) const APP_BUNDLE_IDENTIFIER: &str = "com.recupere.desktop";

pub(crate) fn runtime_capabilities() -> RuntimeCapabilities {
    // `optional_cloud_ai` now reflects whether the bundled Gemma backend is
    // ready (Ollama reachable + model installed). The field name is kept for
    // serialization compatibility with the front-end mappers.
    let gemma_ready = crate::cloud_ai::check_status(&crate::cloud_ai::load_settings()).ready;
    RuntimeCapabilities {
        device_detection: true,
        heuristic_diagnostic: true,
        ai_advisory: true,
        optional_cloud_ai: gemma_ready,
        scan_engine: true,
        imaging_engine: true,
        results_browser: true,
        export_validation: true,
        export_engine: true,
        technical_logs: true,
        limited_capabilities: vec![
            "aiAdvisory".into(),
            "scanEngine".into(),
            "imagingEngine".into(),
            "technicalLogs".into(),
        ],
    }
}

pub(crate) fn app_build_info() -> AppBuildInfo {
    AppBuildInfo {
        product_name: APP_PRODUCT_NAME.into(),
        bundle_identifier: APP_BUNDLE_IDENTIFIER.into(),
        app_version: env!("CARGO_PKG_VERSION").into(),
        package_name: env!("CARGO_PKG_NAME").into(),
        build_profile: if cfg!(debug_assertions) {
            "debug".into()
        } else {
            "release".into()
        },
        operating_system: env::consts::OS.into(),
        architecture: env::consts::ARCH.into(),
        target_triple: option_env!("TARGET").unwrap_or("unknown").into(),
        tauri_runtime: "desktop".into(),
    }
}

#[cfg_attr(feature = "desktop", tauri::command)]
pub fn get_runtime_capabilities() -> RuntimeCapabilities {
    tracing::info!("get_runtime_capabilities: returning available desktop capabilities");
    runtime_capabilities()
}

#[cfg_attr(feature = "desktop", tauri::command)]
pub fn get_app_build_info() -> AppBuildInfo {
    tracing::info!("get_app_build_info: returning build/runtime identity");
    app_build_info()
}
